Add Zamba2

pglorio opened the pull request:
# What does this PR do?

Please include support for Zamba2 architecture created by Zyphra Technologies.

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@ArthurZucker

pglorio committed:
First commit

pglorio committed:
First commit

pglorio committed:
Finish model implementation

pglorio committed:
Finish model implementation

pglorio committed:
Merge branch 'zamba2' of https://github.com/Zyphra/transformers_zamba into zamba2

pglorio committed:
Register zamba2

pglorio committed:
generated modeling and configuration

pglorio committed:
Merge pull request #2 from Zyphra/main

Rebase zamba2

pglorio committed:
generated modeling and configuration

pglorio committed:
added hybrid cache

pglorio committed:
fix attention_mask in mamba

pglorio committed:
dropped unused loras

pglorio committed:
fix flash2

pglorio committed:
Merge pull request #3 from Zyphra/main

rebase

Commit:
config docstrings

pglorio committed:
fix config and fwd pass

pglorio committed:
make fixup fixes

pglorio committed:
text_modeling_zamba2

pglorio committed:
Merge pull request #4 from Zyphra/main

Rebase

pglorio committed:
small fixes

pglorio committed:
make fixup fixes

pglorio committed:
Merge pull request #5 from Zyphra/main

Rebase

pglorio committed:
Fix modular model converter

pglorio commented:
Hey @arthur,

Thank you again for your help in getting Zamba2 into `transformers`! The PR is now finally ready to be reviewed. I added the documentation and all unit tests pass, including slow tests.

A few remarks, mostly related to `modular transformers`:

1. To generate modeling and configuration I used `utils/modular_model_converter.py` from a [previous commit](https://github.com/huggingface/transformers/blob/9e3d704e2340fe9b306b5bd6b12605e4341c012b/utils/modular_model_converter.py) because the most [recent version](https://github.com/huggingface/transformers/blob/e2ac16b28a0b8b900e136750309ca40c49d975c5/utils/modular_model_converter.py) of this script that followed from a large refactoring produces an error that I was not able to fix:
```
Converting src/transformers/models/zamba2/modular_zamba2.py to a single model single file format
Traceback (most recent call last):
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1510, in <module>
    converted_files = convert_modular_file(file_name, args.old_model_name, args.new_model_name)
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1447, in convert_modular_file
    for file, module in create_modules(cst_transformers).items():
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1387, in create_modules
    nodes_to_add, file_type, new_imports = get_class_node_and_dependencies(modular_mapper, class_name, node, files)
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1337, in get_class_node_and_dependencies
    new_node_dependencies, new_imports = check_dependencies_and_create_import_node(
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1283, in check_dependencies_and_create_import_node
    class_dependencies = {dep for dep in new_dependencies if m.matches(mapper.global_nodes[dep], m.ClassDef())}
  File "/workspace/transformers_zamba/utils/modular_model_converter.py", line 1283, in <setcomp>
    class_dependencies = {dep for dep in new_dependencies if m.matches(mapper.global_nodes[dep], m.ClassDef())}
KeyError: 'Zamba2Config'
```

I carefully compared `Zamba2Config` with classes of other models that also use `modular` (such as `Gemma2Config`) and they appear to have consistent format. Relatedly, the `utils/modular_model_converter.py` in the current PR ([path](https://github.com/Zyphra/transformers_zamba/blob/ffc2a58f3dfdee943c8343c73e6155f025de780f/utils/modular_model_converter.py)) is the version from the [previous commit](https://github.com/huggingface/transformers/blob/9e3d704e2340fe9b306b5bd6b12605e4341c012b/utils/modular_model_converter.py) mentioned above.

2. After running `utils/modular_model_converter.py`, the modeling and configuration files generated contain unintended code that I had to update. All these modifications are in [this commit](https://github.com/Zyphra/transformers_zamba/commit/9adf85e00b42a7dca00ff13c8646f504dad6c2c4). In particular, the produced modeling file contains `Zamba2DynamicCache`, which is the correct cache of Zamba2 as well as `HybridMambaAttentionDynamicCache`, which is the cache of Zamba and is not relevant to Zamba2, so I deleted `HybridMambaAttentionDynamicCache` and related references.

3. I ran `make fixup` and all zamba-related tests pass, with the exception of `python utils/check_modular_conversion.py`. This test doesn't pass due to the modifications mentioned in the previous point.

4. I slightly edited the Zamba2MambaMixer compared to the original Mamba2Mixer of mamba2, the main difference is that I added [these lines](https://github.com/Zyphra/transformers_zamba/blob/4da8d5ff6dc5aed96db4d7ec6f2dbf4a3255688c/src/transformers/models/zamba2/modular_zamba2.py#L1138-L1142), which was necessary to appropriately process the mamba2 cache (note this step already existed in the `torch forward` in [these lines](https://github.com/Zyphra/transformers_zamba/blob/4da8d5ff6dc5aed96db4d7ec6f2dbf4a3255688c/src/transformers/models/zamba2/modular_zamba2.py#L1213-L1217)).

Looking forward to your feedback. Thanks so much!

pglorio committed:
added inheritances in modular, renamed zamba cache

pglorio committed:
Merge pull request #6 from Zyphra/main

rebase on upstream

pglorio committed:
modular rebase

pglorio committed:
Rebase

pglorio committed:
new modular conversion

pglorio committed:
fix generated modeling file

pglorio committed:
fixed import for Zamba2RMSNormGated

pglorio committed:
modular file cleanup

pglorio committed:
rebase

pglorio committed:
make fixup and model tests

pglorio committed:
dropped inheritance for Zamba2PreTrainedModel

pglorio committed:
make fixup and unit tests

pglorio commented:
Thank you @ArthurZucker! 

To follow up on your previous message:

- We are using LoRA modules inside the pretrained model. More explicitly, inside the shared transformer blocks, we added un-shared LoRA layers to give some amount of non-shared parameters inside the transformer forward (see e.g. [here](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1232)). We call these layers LoRA because they have the same structure as the usual LoRA, but they are unrelated to finetuning, they are part of the base model and should be included in the modular/modeling files.
- 

> mamba2 should be used for the mamba chunk scan implementation

I may not be understanding this, do you mean to replace [these lines](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L55-L66) with:

```
from ..mamba2.modeling_mamba2 import (
    mamba_chunk_scan_combined, 
    mamba_split_conv1d_scan_combined, 
    selective_state_update, 
    causal_conv1d_fn, 
    causal_conv1d_update,
)
```
? 
Just to be sure: we did not import directly the mamba2 layer (i.e., `mamba2.Mamba2Mixer`) because both the init and the forward methods are slightly different. As mentioned in a [previous msg](https://github.com/huggingface/transformers/pull/34517#issuecomment-2467369443), we made a few edits inside of the forward of this layer, mostly related to the `attention_mask`. In the init, some of the attributes are different, e.g. we set `self.num_heads = self.config.n_mamba_heads` instead of `self.num_heads = config.num_heads` because in the config we distinguish between `num_attention_heads` and `n_mamba_heads`, while `num_heads` would be ambiguous.


Also, below are the updates I made:

- I inherited the `Zamba2HybridDynamicCache` from `ZambaHybridDynamicCache` as you requested. In the `__init__` method, I omitted `super().__init__()` as discussed with @Cyrilvallez, because this would add unwanted lines involving `self.conv_states` and `self.ssm_states` that come from `ZambaHybridDynamicCache` (`Zamba2HybridDynamicCache` manages the SSM cache the same way as mamba2, which uses a slightly different structure for `self.conv_states` and `self.ssm_states` compared to `ZambaHybridDynamicCache`).
- Regarding the config, I didn't inherit it from `ZambaConfig` because some flags of `ZambaConfig` do not exist in `Zamba2Config`, such as `use_mamba_kernels`. Also many default values of `Zamba2Config` are different from those of `ZambaConfig`. I noticed that Gemma2Config doesn't inherit from the GemmaConfig, so because of these reasons I thought it is a simpler option to just define the config from scratch. We're happy to revisit this if needed.
- I imported `MambaRMSNormGated` from mamba2
- Inherited `Zamba2AttentionDecoderLayer` (I overrode the forward method because `hidden_states = self.feed_forward(hidden_states, layer_idx)` takes the `layer_idx` in [Zamba2](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1343), this was not the case for [Zamba](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba/modeling_zamba.py#L847))
- Inherited `Zamba2MLP`. There are three differences with `ZambaMLP`: 1. added [pretrained LoRA](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1251) (as described above) and 2. replaced `self.up_proj`, and `self.gate_proj` with their [concatenation](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1248) `self.gate_up_proj` (this is necessary as the pretrained LoRA was added collectively to the concat of up and gate), 3. modified the [activation](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1243) as it should take the output of self.gate_up_proj as input, instead of the output of self.gate_proj.
- I renamed shared_transf -> shared_transformer
- In `Zamba2HybridLayer`, I [overrode the forward](https://github.com/Zyphra/transformers_zamba/blob/10a0b1e1bcc8d8b541ea93b9bddeafc20e695858/src/transformers/models/zamba2/modular_zamba2.py#L1371) of `ZambaHybridLayer` because the latter has `self.shared_transf` in its forward method (I cannot change `shared_transf` -> `shared_transformer` in Zamba1 because this would require replacing ~200 checkpoints stored in the hub). The only way to avoid overriding the forward would be to keep `shared_transf` instead of renaming it to `shared_transformer`.
- As agreed with @Cyrilvallez, I did not inherit `Zamba2PreTrainedModel`. The reason is that if I inherit `Zamba2PreTrainedModel` I need to overload <this method> in `ZambaPreTrainedModel` with <this method> in `PretrainedModel`. Calling `super(PreTrainedModel, cls)._check_and_enable_flash_attn_2` didn't work, so I found it easier to simply not inherit `Zamba2PreTrainedModel` (in fact, this approach makes the class in modular shorter).
- I re-ran `make fixup`, all tests pass except two that are unrelated to zamba:
```
python utils/check_modular_conversion.py

Differences found between the generated code and src/transformers/models/gemma/configuration_gemma.py:

   1 --- src/transformers/models/gemma/configuration_gemma.py_generated
   2 +++ src/transformers/models/gemma/configuration_gemma.py
   3 @@ -19,6 +19,7 @@
   4  # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   5  # See the License for the specific language governing permissions and
   6  # limitations under the License.
   7 +
   8  
   9  from ...configuration_utils import PretrainedConfig
  10  
Traceback (most recent call last):
  File "/workspace/transformers_zamba/utils/check_modular_conversion.py", line 78, in <module>
    raise ValueError("Some diff and their modeling code did not match.")
ValueError: Some diff and their modeling code did not match.
```
and
```
Traceback (most recent call last):
  File "/workspace/transformers_zamba/utils/check_repo.py", line 1211, in <module>
    check_repo_quality()
  File "/workspace/transformers_zamba/utils/check_repo.py", line 1195, in check_repo_quality
    check_all_objects_are_documented()
  File "/workspace/transformers_zamba/utils/check_repo.py", line 1076, in check_all_objects_are_documented
    check_public_method_exists(documented_methods_map)
  File "/workspace/transformers_zamba/utils/check_repo.py", line 1125, in check_public_method_exists
    raise Exception("\n".join(failures))
Exception: The following public method is explicitly documented but not defined in the corresponding class. class: MusicgenMelodyFeatureExtractor, method: _extract_stem_indices. If the method is defined, this error can be due to lacking dependencies.
```


Thank you!

Cyrilvallez commented:
Hey @pglorio, so sorry for the delay!! I think the first error you encounter comes from a wrong `ruff` version. Our CI uses `ruff==0.5.1`, so could you make sure to install the same and run it again?

Cyrilvallez reviewed (commented):
Hey! Here is a first round of reviews, you can check out the comments! For the lora layers, maybe rename them `residual` instead of `lora` to make it explicit that it is part of the original model?

You should also make sure that all the tests you added pass, this is not the case actually!

pglorio committed:
Add inheritance of rope from GemmaRotaryEmbedding

pglorio committed:
moved rope to model init

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+                self.head_dim,
+                max_position_embeddings=config.max_position_embeddings,
+                base=self.rope_theta,
+            )
```
I moved rope to model init ([commit](https://github.com/huggingface/transformers/pull/34517/commits/8079ae035614cdf662af92093c3a84887639821a)). 

As a remark, I overrode the forward pass of `ZambaModel` in `Zamba2Model`, since Zamba doesn't have rope. Similarly, I overrode the forward pass of `ZambaMambaDecoderLayer` in `Zamba2MambaDecoderLayer` because the former doesn't take `position_embeddings` as argument (and it wouldn't make sense to add it since there is no rope in Zamba).

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+
+        if self.config.use_mem_rope:
+            cos, sin = self.rotary_emb(value_states, position_ids)
+            query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
```
done!

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+    "The bare Zamba2 Model outputting raw hidden-states without any specific head on top.",
+    ZAMBA2_START_DOCSTRING,
+)
+class Zamba2Model(ZambaModel, Zamba2PreTrainedModel):
+    """
+    Model consisting of *config.num_hidden_layers* layers.
+
```
done!

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+        return out
+
+    # fmt: off
+    def torch_forward(self, input_states, cache_params: Optional[Zamba2HybridDynamicCache]=None, cache_position:Optional[torch.LongTensor]=None, attention_mask: Optional[torch.Tensor]=None):
+        batch_size, seq_len, _ = input_states.shape
+        dtype = input_states.dtype
+        # Gated MLP's linear projection
```
This function is adapted from Mamba2, with the slight modification of the if statement implementing the attention mask ([zamba2 if statement](https://github.com/Zyphra/transformers_zamba/blob/zamba2/src/transformers/models/zamba2/modular_zamba2.py#L1062), [mamba2 if statemet](https://github.com/Zyphra/transformers_zamba/blob/zamba2/src/transformers/models/mamba2/modeling_mamba2.py#L439)).

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+                out = self.out_proj(scan_output)
+        return out
+
+    # fmt: off
```
We adapted this class from `Mamba2Mixer` in Mamba2 (up to small modifications mentioned in point 4 of [this message](https://github.com/huggingface/transformers/pull/34517#issuecomment-2467369443)), which also removed formatting https://github.com/huggingface/transformers/blob/main/src/transformers/models/mamba2/modeling_mamba2.py#L406

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+        super().__init__(config, layer_idx)
+        num_gs = count_mem_blocks_in_config(config)
+        self.block_id = block_id
+        del self.self_attn
+        del self.feed_forward
```
Thanks, dropped these!

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+class Zamba2AttentionDecoderLayer(ZambaAttentionDecoderLayer):
+    def __init__(self, config: Zamba2Config, block_id: int = None, layer_idx: Optional[int] = None):
+        super().__init__(config, layer_idx)
+        num_gs = count_mem_blocks_in_config(config)
+        self.block_id = block_id
```
Done, thanks for catching it!

pglorio committed:
drop del self.self_attn and del self.feed_forward

pglorio committed:
Rebase onto upstream

pglorio committed:
fix tests

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+        self.is_causal = True
+
+        if config.use_shared_attention_lora:
+            self.linear_q_lora_A_list = nn.ModuleList([])
+            self.linear_q_lora_B_list = nn.ModuleList([])
+            self.linear_k_lora_A_list = nn.ModuleList([])
+            self.linear_k_lora_B_list = nn.ModuleList([])
+            self.linear_v_lora_A_list = nn.ModuleList([])
+            self.linear_v_lora_B_list = nn.ModuleList([])
```
We'd be happy to update this if strictly necessary, although this would also involve converting all the checkpoints on the hub. The use of `nn.Sequential` will inherently change the weight keys as they currently reflect the separation into `A` and `B` layers. For example: `linear_q_lora_A_list.<index>.weight` would change to something like `linear_q_lora_list.<index>.<index>.weight`. Please let us know if there is still a need for this to be modified.

pglorio replied:
done as you suggested [in this commit](https://github.com/huggingface/transformers/pull/34517/commits/f1b31a13289255e55f612041dea9df4a467831e6#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a)

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+
+            for i in range(self.num_fwd_mem_blocks):
+                if i % config.num_mem_blocks == block_id:
+                    linear_q_lora_A = nn.Linear(self.attention_hidden_size, self.config.lora_rank, bias=False)
+                    linear_q_lora_B = nn.Linear(self.config.lora_rank, self.attention_hidden_size, bias=False)
```
See above comment.

pglorio replied:
done as you suggested [here](https://github.com/huggingface/transformers/pull/34517/commits/f1b31a13289255e55f612041dea9df4a467831e6#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a)

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        bsz, q_len, _ = hidden_states.size()
+
+        if self.config.use_shared_attention_lora:
+            lora_layer_idx = self.layer_dic[layer_idx]
+            linear_q_lora_A = self.linear_q_lora_A_list[lora_layer_idx]
+            linear_q_lora_B = self.linear_q_lora_B_list[lora_layer_idx]
+            q_lora_output = linear_q_lora_A(hidden_states)
+            q_lora_output = linear_q_lora_B(q_lora_output)
+            query_states = self.q_proj(hidden_states)
+            query_states = query_states + q_lora_output
+            linear_k_lora_A = self.linear_k_lora_A_list[lora_layer_idx]
+            linear_k_lora_B = self.linear_k_lora_B_list[lora_layer_idx]
+            k_lora_output = linear_k_lora_A(hidden_states)
+            k_lora_output = linear_k_lora_B(k_lora_output)
+            key_states = self.k_proj(hidden_states)
+            key_states = key_states + k_lora_output
+            linear_v_lora_A = self.linear_v_lora_A_list[lora_layer_idx]
+            linear_v_lora_B = self.linear_v_lora_B_list[lora_layer_idx]
+            v_lora_output = linear_v_lora_A(hidden_states)
+            v_lora_output = linear_v_lora_B(v_lora_output)
+            value_states = self.v_proj(hidden_states)
+            value_states = value_states + v_lora_output
+        else:
+            query_states = self.q_proj(hidden_states)
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
```
Thank you for outlining this option. As discussed above, the use of `nn.Sequential` would require renaming weight keys in several base and instruct checkpoints that we uploaded on the hub.

pglorio replied:
We have updated the code following your suggestions by including `nn.Sequential` in [this commit](https://github.com/huggingface/transformers/pull/34517/commits/f1b31a13289255e55f612041dea9df4a467831e6#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a).

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1760 @@
+        self.gate_up_proj = nn.Linear(self.hidden_size, 2 * self.intermediate_size, bias=config.add_bias_linear)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.add_bias_linear)
+
+        if self.config.use_shared_mlp_lora:
+            self.gate_up_proj_lora_A_list = nn.ModuleList([])
+            self.gate_up_proj_lora_B_list = nn.ModuleList([])
+            for i in range(self.num_fwd_mem_blocks):
+                if i % config.num_mem_blocks == block_id:
+                    gate_up_proj_lora_A = nn.Linear(self.config.hidden_size, self.config.lora_rank, bias=False)
+                    gate_up_proj_lora_B = nn.Linear(self.config.lora_rank, 2 * self.intermediate_size, bias=False)
+                else:
+                    gate_up_proj_lora_A = nn.Identity()
+                    gate_up_proj_lora_B = nn.Identity()
+                self.gate_up_proj_lora_A_list.append(gate_up_proj_lora_A)
+                self.gate_up_proj_lora_B_list.append(gate_up_proj_lora_B)
+
+        layer_block_map = layer_type_list(config)
+        self.layer_dic = {value: index for index, value in enumerate(layer_block_map)}
+
+    def forward(self, hidden_state, layer_idx=None):
+        if self.config.use_shared_mlp_lora:
+            layer_idx = self.layer_dic[layer_idx]
+            gate_up_proj_lora_A = self.gate_up_proj_lora_A_list[layer_idx]
+            gate_up_proj_lora_B = self.gate_up_proj_lora_B_list[layer_idx]
+            lora_output = gate_up_proj_lora_A(hidden_state)
+            lora_output = gate_up_proj_lora_B(lora_output)
+            intermediate_state = self.gate_up_proj(hidden_state)
+            hidden_state = intermediate_state + lora_output
+        else:
+            hidden_state = self.gate_up_proj(hidden_state)
```
See above comment.

pglorio replied:
done as you suggested, see above

pglorio commented:
Thank you @cyrilv for the super helpful feedback! Please see above the replies to each comment.

It seems zamba2-related tests all pass locally for us, but not in the circle/ci setting of the PR. The error is always 
```
   1 failed because `IndexError` -> Dimension out of range (expected to be in range of [-2, 1], but got 2)
   9 failed because `RuntimeError` -> Sizes of tensors must match except in dimension 2. Expected size 8 but got size 1 for tensor number 1 in the list.
```
for all tests that fail. We investigated the discrepancy by examinating more closely the test `tests/models/zamba2/test_modeling_zamba2.py::Zamba2ModelTest::test_beam_sample_generate`. We tried using `pytest-flakefinder` as suggested [here](https://github.com/huggingface/transformers/pull/26304#issuecomment-1819382300) and using this command `python3 -m pytest -m 'generate' -n 8 -vvv --dist=loadfile --flake-finder --flake-runs=1000 --max-worker-restart=0 tests/models/zamba2/test_modeling_zamba2.py::Zamba2ModelTest::test_beam_sample_generate` but the test still passes. This is the closest reproduction of the circle/ci test that we could reproduce to the best of our knowledge. We'd appreciate if we can get a hint of what we can do to fix this discrepancy!

Also, we are happy to rename lora layers as you suggested [here](https://github.com/huggingface/transformers/pull/34517#pullrequestreview-2476481472) and agree it would clarify the role of this module. Would `pretrained_lora` be also good as an option (`residual` might for some people refer to the identity stream)?

Thank you so much again!

pglorio committed:
renamed lora -> adapter

pglorio committed:
rewrote adapter implementation

pglorio committed:
rebase

pglorio committed:
fixed tests

pglorio commented:
We renamed lora -> adapter [here](https://github.com/huggingface/transformers/pull/34517/commits/337faed6afa70c244357c44b2076bd3920728d04) and, following your suggestion in [this comment](https://github.com/huggingface/transformers/pull/34517#discussion_r1868257901) we simplified the adapter using `nn.Sequential` in [this commit](https://github.com/huggingface/transformers/pull/34517/commits/f1b31a13289255e55f612041dea9df4a467831e6).

We re-ran the model tests and they all pass locally, but it looks like they still fail in this PR environment.

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1872 @@
+                self.linear_q_adapter_list.append(linear_q_adapter)
+                self.linear_k_adapter_list.append(linear_k_adapter)
+                self.linear_v_adapter_list.append(linear_v_adapter)
+
+        if config.use_mem_rope:
+            rope_theta = config.rope_theta
+            if config.use_long_context:
+                a = 8
+                rope_theta = rope_theta * a ** (config.attention_head_dim / (config.attention_head_dim - 2))
+            self.rotary_emb = Zamba2RotaryEmbedding(
+                config.attention_head_dim,
+                max_position_embeddings=config.max_position_embeddings,
+                base=rope_theta,
+            )
+
+        self.layer_dic = {value: index for index, value in enumerate(self.layer_block_map)}
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        layer_idx: int,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Zamba2HybridDynamicCache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # will become mandatory in v4.46
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        bsz, q_len, _ = hidden_states.size()
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+        if self.config.use_shared_attention_adapter:
+            adapter_layer_idx = self.layer_dic[layer_idx]
+            query_states += self.linear_q_adapter_list[adapter_layer_idx](hidden_states)
+            key_states += self.linear_k_adapter_list[adapter_layer_idx](hidden_states)
+            value_states += self.linear_v_adapter_list[adapter_layer_idx](hidden_states)
```
nice, this is so much simpler 🤗

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1872 @@
+                    if config.use_shared_attention_adapter:
+                        tied_keys_adapter = []
+                        adapter_id = 0
+                        for _layer_type in self.layers_block_type:
+                            if _layer_type == "hybrid" and adapter_id % config.num_mem_blocks == block.block_id:
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_q_adapter_list."
+                                    + str(adapter_id)
+                                    + ".0.weight"
+                                )
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_k_adapter_list."
+                                    + str(adapter_id)
+                                    + ".0.weight"
+                                )
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_v_adapter_list."
+                                    + str(adapter_id)
+                                    + ".0.weight"
+                                )
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_q_adapter_list."
+                                    + str(adapter_id)
+                                    + ".1.weight"
+                                )
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_k_adapter_list."
+                                    + str(adapter_id)
+                                    + ".1.weight"
+                                )
+                                tied_keys_adapter.append(
+                                    "shared_transformer.self_attn.linear_v_adapter_list."
+                                    + str(adapter_id)
+                                    + ".1.weight"
+                                )
+                            adapter_id += 1
+                        self._tied_weights_keys = [*self._tied_weights_keys, *tied_keys_adapter]
+                layers.append(Zamba2HybridLayer(block, next(linear_layers), next(mamba_layers)))
+            else:
+                layers.append(next(mamba_layers))
```
Putting this block in another method such as `get_layers` would help lisibility a great deal!

pglorio replied:
Done [here](https://github.com/huggingface/transformers/pull/34517/commits/095d853bf872e9d05036dddf4fd8f011cb8646dd).

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1872 @@
+                self.linear_k_adapter_list.append(linear_k_adapter)
+                self.linear_v_adapter_list.append(linear_v_adapter)
+
+        if config.use_mem_rope:
+            rope_theta = config.rope_theta
+            if config.use_long_context:
+                a = 8
+                rope_theta = rope_theta * a ** (config.attention_head_dim / (config.attention_head_dim - 2))
+            self.rotary_emb = Zamba2RotaryEmbedding(
+                config.attention_head_dim,
+                max_position_embeddings=config.max_position_embeddings,
+                base=rope_theta,
+            )
```
This should be removed (defined in the Model), and `position_embeddings` should be passed everywhere as a required arg (not a keyword arg)!

pglorio replied:
Removed [here](https://github.com/huggingface/transformers/pull/34517/commits/99708af8505771e07d77cee20fcc7c533e562861). We cannot make `position_embeddings` required as some zamba2 models do not have rope.

I also updated rope [here](https://github.com/huggingface/transformers/pull/34517/commits/d9b4a5004af2f53e281038c66ee4f984bb53dafe#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL382). I modified the init method compared to the inherited rope as we need to pass custom `dim` and `base` arguments.

Cyrilvallez reviewed (commented):
Nice! I pointed out some minor issues, but this is almost ready! 🤗 However, you might want to consider switching the attention implementation to our newest standard here https://github.com/huggingface/transformers/pull/35235 (see e.g. the llama model as an example). This is not strictly needed, but should be quite straightforward from the examples, and will help both lisibility and maintenance in the long run. And as this change was introduced to make transformers compatible with high-performance inference frameworks such as vLLM and TGI, it would potentially make Zamba2 compatible as well (not sure how we will support stateful models, and if the shapes in the Mixers are compatible, but cannot hurt!) in the future!

Cyrilvallez commented:
I cannot reproduce your issue with tests though, they are failing locally for me. Running `CUDA_VISIBLE_DEVICES=2 pytest tests/models/zamba2/test_modeling_zamba2.py` gives:
<img width="1340" alt="Screenshot 2024-12-18 at 17 47 59" src="https://github.com/user-attachments/assets/f6d00761-acfb-4dd1-84fc-0f0431a3911a" />

pglorio committed:
Merge branch 'main' into zamba2

pglorio committed:
Fix torch_forward in mamba2 layer

pglorio committed:
Fix torch_forward in mamba2 layer

pglorio committed:
Fix torch_forward in mamba2 layer

pglorio committed:
Dropped adapter in-place sum

pglorio committed:
removed rope from attention init

pglorio committed:
updated rope

pglorio committed:
created get_layers method

pglorio commented:
Thank you so much for [this feedback](https://github.com/huggingface/transformers/pull/34517#issuecomment-2551816849) @Cyrilvallez. I realized that the issue with unit tests was inside the `torch_forward` method of the mamba2 mixer (when i ran locally, the unit tests used `cuda_kernels` method instead). I fixed that method here: [1](https://github.com/huggingface/transformers/pull/34517/commits/5d0a5d46c79f83b97b5bf24eb42a9c47750decce) [2](https://github.com/huggingface/transformers/pull/34517/commits/ef055c90de487771d5ee70f9fac148c752c99c96) [3](https://github.com/huggingface/transformers/pull/34517/commits/b993a7895e7e42be6a3f9647759a19320f6bd73e).

By the way, we originally took the the `torch_forward` from the mamba2 model, so the same issues hold there. In particular, running this:

```python
config = Mamba2Config(num_heads=8,
        n_groups=8,
        state_size=2,
        head_dim=8,
        conv_kernel=4,
        chunk_size=8,
        vocab_size=99,
        hidden_size=32,
        num_hidden_layers=4,
        hidden_act="silu",
        hidden_dropout_prob=0.1,
        max_position_embeddings=512,
                      )
model = Mamba2ForCausalLM(config)

inputs = {'input_ids': torch.tensor([[86,  6, 51,  3, 12, 15, 33, 18,  4, 92],
         [69, 66, 49, 45, 48, 44, 61, 56, 68, 85]]),
 'attention_mask': torch.tensor([[0, 0, 1, 1, 1, 0, 0, 0, 1, 0],
         [0, 1, 1, 0, 1, 1, 1, 0, 1, 1]])}

outputs_cpu = model.generate(**inputs, max_new_tokens=5, return_dict_in_generate=False, output_scores=False, use_cache=True, num_beams=1, do_sample=False)
model = model.to('cuda')
inputs = {key: tensor.to(device=model.device) for key, tensor in inputs.items()}
outputs_cuda = model.generate(**inputs, max_new_tokens=5, return_dict_in_generate=False, output_scores=False, use_cache=True, num_beams=1, do_sample=False)
print(torch.all(outputs_cpu == outputs_cuda.cpu()).item())
```
returns `False`.

pglorio committed:
rebase

pglorio committed:
make fixup fix

pglorio committed:
make fixup fixes

pglorio committed:
make fixup fixes

pglorio committed:
fix merge conflicts

pglorio committed:
update to new attention standard

pglorio committed:
fixes for merge

pglorio committed:
update to new attention standard

pglorio committed:
make fixup fixes

pglorio commented:
Hi @Cyrilvallez and @ArthurZucker,

I updated attention to the new standard of `transformers` [here](https://github.com/huggingface/transformers/pull/34517/commits/cb90bb4efbab0243bc86ada0203dacfb60080c8b) and [here](https://github.com/huggingface/transformers/pull/34517/commits/1dbc8c73e54e71345cfe676eea704aa862b7fc84). 

I ran all final tests, including `@slow` tests, and everything appears to pass!

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+    is_causal_conv1d_available,
+    is_mamba_ssm_available,
+)
+from ..gemma.modeling_gemma import GemmaRotaryEmbedding
+
+# from ..llama.modeling_llama import apply_rotary_pos_emb
```
Small detail but let's import the Rotary from Llama as a reference instead of Gemma

pglorio replied:
[done](https://github.com/huggingface/transformers/blob/2b29338bd2c4f1becb36caa65199e5d398c2b2a0/src/transformers/models/zamba2/modular_zamba2.py#L38)!

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+logger = logging.get_logger(__name__)
+
+
+class Zamba2Config(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`Zamba2Model`]. It is used to instantiate a
+    Zamba2 model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the Zamba2 model.
```
If we are not inheriting the Config using modular, I'd rather you put it directly into `configuration_zamba2` for clarity! (it clutters a bit the modular without good reason otherwise)

pglorio replied:
How do i create configuration_zamba2.py? I tried to manually move the configuration from modular _zamba2.py to that file but when I run the converter it overwrites configuration_zamba2.py and it produces an almost empty file where the only non-commented-out content is `__all__ = ["Zamba2Config"]`.

pglorio replied:
thanks! done [here](https://github.com/huggingface/transformers/pull/34517/commits/5ace701e4c7e5931d459de0f96d2c3c46291ccda) and [here](https://github.com/huggingface/transformers/pull/34517/commits/535b6319966b3ce53525cb86e8df0f6971bac464)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+    return output_list
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
+    Args:
+        q (`torch.Tensor`): The query tensor.
+        k (`torch.Tensor`): The key tensor.
+        cos (`torch.Tensor`): The cosine part of the rotary embedding.
+        sin (`torch.Tensor`): The sine part of the rotary embedding.
+        position_ids (`torch.Tensor`, *optional*):
+            Deprecated and unused.
+        unsqueeze_dim (`int`, *optional*, defaults to 1):
+            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
+            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
+            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
+            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
+            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
+            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+    Returns:
+        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
```
https://github.com/huggingface/transformers/pull/35692 will allow you to import those directly! I'll try to merge it very soon so that you can use it 🤗

pglorio replied:
Sounds good!

Cyrilvallez replied:
Just merged it!

pglorio replied:
awesome! applied it [here](https://github.com/huggingface/transformers/pull/34517/commits/1c92266d3735ff2c8ce9cbc64dab5a7db8434e49)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+
+
+class Zamba2RotaryEmbedding(GemmaRotaryEmbedding):
+    def __init__(
+        self,
+        config: Zamba2Config,
+        device=None,
+    ):
+        self.rope_kwargs = {"base": config.rope_theta, "dim": config.attention_head_dim}
+        super().__init__(config, device)
+        inv_freq, self.attention_scaling = self.rope_init_fn(config=None, device=device, **self.rope_kwargs)
```
It should not be needed to add the `rope_kwargs`, and we actually removed them from the call recently 😉 
You can just `pass` the definition and it will use the Config

pglorio replied:
Unfortunately we need to pass `dim` explicitly and `self.rope_kwards` is to the best of my understanding the only way to make it work right now. If we don't do it this way, [this line](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py#L61) computes `dim` as `getattr(config, "head_dim", config.hidden_size // config.num_attention_heads)`. In Zamba2, `head_dim` is `2 * config.hidden_size // config.num_attention_heads`, so the only option here would be to set `config.head_dim` explicitly in the config. However, this would be confusing as in Zamba2 we have both mamba heads and attention heads, and thus we prefer avoiding adding the attribute `head_dim` to the config which would be too vague to interpret.

Cyrilvallez replied:
Oh I see, did not notice the factor `2` sorry. Then let's do
```py
def __init__(
        self,
        config: Zamba2Config,
        device=None,
    ):
        super().__init__(config, device)
       # we cannot use the config here to parameterize because of a factor 2 for the head_dim
        inv_freq, self.attention_scaling = self.rope_init_fn(device=device, base=config.rope_theta, dim=config.attention_head_dim)
```
it's just slightly cleaner

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/99bde9383fdcaa1abb156be0fc2ea440c6d8c59f)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        past_key_value: Optional[Zamba2HybridDynamicCache] = None,
+        use_cache: bool = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # will become mandatory in v4.46
```
You can make them mandatory here! Let's remove the default value for clarity

pglorio replied:
`position_embeddings` can be `None` depending on the model: some zamba2 models (specifically, `Zyphra/Zamba2-2.7B`) don't have rope (in which case `config.use_mem_rope=False`) so `position_embeddings` should be `None`. Thus, for zamba2 this argument should be optional. I removed the inline comment [here](https://github.com/huggingface/transformers/pull/34517/commits/1e3b51e5c08283a1a4e954506f8d07d27bb07f9a#diff-10003c7c32bb1e17589aea478bcf05e9d0cde6f0f2770cb768874966398d4297L451-R448).

Separately, I removed a few occurrencies of `position_ids`, `use_cache` and `cache_position` from signatures of methods in which they are not used [here](https://github.com/huggingface/transformers/pull/34517/commits/1e3b51e5c08283a1a4e954506f8d07d27bb07f9a).

Cyrilvallez replied:
Perfect, works perfectly for me then!

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        layer_idx: int,
```
I did not fully check the logic here, but do we really need the `layer_idx` in the `forward`'s signature of all modules? If we are not mixing them, put them into the `__init__` and use `self.layer_idx` instead!

pglorio replied:
This is a shared layer, so yes, `layer_idx` should be specified in the `forward`'s signature rather than in the `__init__`.

Cyrilvallez replied:
Ok, leaving it then!

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        value_states = value_states.view(hidden_shape).transpose(1, 2)
+
+        if self.config.use_mem_rope:
+            if position_embeddings is None:
+                logger.warning_once(
+                    "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
+                    "through `position_ids` (2D tensor with the indexes of the tokens), to using externally computed "
+                    "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.46 `position_ids` will be "
+                    "removed and `position_embeddings` will be mandatory."
+                )
+                cos, sin = self.rotary_emb(value_states, position_ids)
+            else:
```
To remove! We don't have a RotaryLayer here and we want to make the position_embedding necessary

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL517-R517)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        is tied, un-tied adapter modules (formally same as LoRA, but used in the base model) are added to the up and gate projectors to increase expressivity with a small memory overhead.
+        """
+        super().__init__(config)
+        self.config = config
```
No need to rewrite the config here!

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL994)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        self.num_fwd_mem_blocks = num_fwd_mem_blocks
+        self.block_id = block_id
+
+        def gated_act_fn(x):
+            x = torch.chunk(x, 2, dim=-1)
+            return self.act_fn(x[0]) * x[1]
+
+        self.gated_act_fn = gated_act_fn
```
Not a fan of this at all! Let's do it explicitly in the `forward`!

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL998-L1003) and [here](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL1031-R1015)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+
+        self.gated_act_fn = gated_act_fn
+
+        del self.gate_proj
+        del self.up_proj
+        del self.down_proj
+        self.gate_up_proj = nn.Linear(self.hidden_size, 2 * self.intermediate_size, bias=config.add_bias_linear)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.add_bias_linear)
```
Overall might be easier to not inherit this module, but works either way for me! If you keep it like that, please don't `del` attributes that you then re-add! Simply redefine them without the `del` statement before

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL1006)

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[torch.LongTensor] = None,
```
Let's remove the default value of the `position_embeddings` here as well (and all other places that I may have overlooked)

pglorio replied:
see above

Cyrilvallez commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["Zamba2AttentionDecoderLayer", "Zamba2MambaDecoderLayer"]
+    _skip_keys_device_placement = "past_key_values"
+    _supports_flash_attn_2 = True
+    _supports_sdpa = False
```
You can add `_supports_flex_attn` as well!

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aR1273)

Cyrilvallez commented on tests/models/zamba2/test_modeling_zamba2.py:
```diff
@@ -0,0 +1,645 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_for_casual_lm(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_causal_lm(*config_and_inputs)
```
typo in the function name here

pglorio replied:
[fixed](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-67d87c36da1310eb4bb253eb029c63484c1a052ae06224eb5e69e72755ec46b2)

Cyrilvallez commented on tests/models/zamba2/test_modeling_zamba2.py:
```diff
@@ -0,0 +1,645 @@
+                            msg=f"Parameter {name} of model {model_class} seems not properly initialized",
+                        )
+
+    def test_mismatched_shapes_have_properly_initialized_weights(self):
+        r"""
+        Overriding the test_mismatched_shapes_have_properly_initialized_weights test because A_log and D params of the
+        Mamba block are initialized differently and we tested that in test_initialization
+        """
+        self.skipTest("Cumbersome and redundant for Zamba2")
```
We prefer to use the decorator `unittest.skip` for style and understanding faster!

pglorio replied:
[fixed](https://github.com/huggingface/transformers/pull/34517/commits/2b29338bd2c4f1becb36caa65199e5d398c2b2a0#diff-67d87c36da1310eb4bb253eb029c63484c1a052ae06224eb5e69e72755ec46b2)

Cyrilvallez commented on tests/models/zamba2/test_modeling_zamba2.py:
```diff
@@ -0,0 +1,645 @@
+               -6.5103, -6.5104, -6.5106, -6.5105,  7.8700, 13.5434, -6.5104, -6.5096,
+               -6.5106, -6.5102, -6.5106, -6.5106, -6.5105, -6.5106, -6.5104, -6.5106,
+               -6.5105, -6.5106, -6.5106, -6.5113, -6.5102, -6.5105, -6.5108, -6.5105,
+               -6.5104, -6.5106, -6.5106, -6.5104, -6.5106, -6.5107, -6.5103, -6.5105          ]
+            , dtype=torch.float32)  # fmt: skip
+
+        torch.testing.assert_close(logits[0, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_0, rtol=1e-3, atol=1e-3)
+        torch.testing.assert_close(logits[1, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_1, rtol=1e-3, atol=1e-3)
```
Cannot comment on it, but there is a file `utils/modular_model_converter.textClipping` to remove!!

pglorio replied:
removed that file, thanks for spotting it!

Cyrilvallez reviewed (commented):
Nice work for the refactor! Almost ready, left some final comments but overall quite nice! 🤗

pglorio committed:
rebase

pglorio committed:
minor fixes

pglorio committed:
cache_position

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+            )
+
+        if cache_position is None:
+            cache_position = torch.arange(hidden_states.shape[1], device=hidden_states.device)
```
[done](https://github.com/huggingface/transformers/pull/34517/commits/b212cb28cd2430b01580237c7fdacb48a7a516d4#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a)

pglorio committed:
removed cache_position postion_ids use_cache

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[torch.LongTensor] = None,
```
see above

pglorio commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1665 @@
+        # original_hidden_states: word embedding output that will be concatenated with hidden activations to form the input of the shared transformer layer
+
+        if use_cache and past_key_values is None:
+            logger.warning_once(
+                "Zamba2 requires an initialized `Zamba2HybridDynamicCache` to return a cache. None was "
+                "provided, so no cache will be returned."
+            )
+
```
I followed the logic of Zamba, where the cache is instantiated [here](https://github.com/huggingface/transformers/blob/1e3b51e5c08283a1a4e954506f8d07d27bb07f9a/src/transformers/models/zamba/modeling_zamba.py#L1310). This convention is also followed by other mamba-related models such as Jamba and Bamba.

Cyrilvallez replied:
Yes, but without `generate`, `prepare_inputs_for_generation` is never called, so we usually also set it in the Model directly as a fallback if `use_cache` is True 🤗 so let's add it even if not present in previous models!

pglorio replied:
Done [here](https://github.com/huggingface/transformers/pull/34517/commits/baf2ed3f86a0d841e493ab023c210b7124b0f4de).

I had to skip [this test](https://github.com/huggingface/transformers/pull/34517/commits/baf2ed3f86a0d841e493ab023c210b7124b0f4de#diff-67d87c36da1310eb4bb253eb029c63484c1a052ae06224eb5e69e72755ec46b2R328) explicitly as Zamba2 uses a hybrid cache (see docstring there for more details). Before updating the cache instantiation as above, this test was automatically skipped.

pglorio replied:
I also added [this](https://github.com/huggingface/transformers/pull/34517/commits/9afb57ec0dd2e877c7977a188caaaa0b64dc3115#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aR170-R177) method that overrides the one in `ZambaHybridDynamicCache` ([here](https://github.com/huggingface/transformers/blob/9afb57ec0dd2e877c7977a188caaaa0b64dc3115/src/transformers/models/zamba/modeling_zamba.py#L194)). I changed [this line](https://github.com/huggingface/transformers/blob/9afb57ec0dd2e877c7977a188caaaa0b64dc3115/src/transformers/models/zamba2/modular_zamba2.py#L174) so that it correctly computes the sequence length (without this, a few `@slow` tests were failing, now everything works).

pglorio commented:
Thank you @Cyrilvallez for the review. I addressed the comments above, although there are a couple of pending points.

All zamba-related tests appear to pass.

pglorio commented on src/transformers/testing_utils.py:
```diff
@@ -1413,6 +1413,7 @@ def set_model_tester_for_less_flaky_test(test_case):
     # TODO (if possible): Avoid exceptional cases
     exceptional_classes = [
         "ZambaModelTester",
+        "Zamba2ModelTester",
         "RwkvModelTester",
```
@ydshieh for context, when running this test the config of the model is forced to have `num_hidden_layers=1` but other parameters of the config are not updated accordingly so when the model is initialized it errors out as these params are not consistently updated. It's probably also the reason why Zamba was added to this list I imagine.

pglorio committed:
remove config from modular

pglorio committed:
removed config from modular (2)

pglorio committed:
rebase

pglorio committed:
import apply_rotary_pos_emb from llama

pglorio committed:
fixed rope_kwargs

pglorio committed:
Instantiate cache in Zamba2Model

pglorio committed:
fix cache

pglorio committed:
fix @slow decorator

pglorio commented:
Hello @Cyrilvallez, I ran all model tests on two GPUs and after a couple of minor fixes everything appears to work now. I'm skipping [this](https://github.com/huggingface/transformers/pull/34517/commits/9afb57ec0dd2e877c7977a188caaaa0b64dc3115#diff-67d87c36da1310eb4bb253eb029c63484c1a052ae06224eb5e69e72755ec46b2R336-R339) test as it gives an error related to mamba2 kernels. I indeed verified that mamba2 skips that test [here](https://github.com/huggingface/transformers/blob/9afb57ec0dd2e877c7977a188caaaa0b64dc3115/tests/models/mamba2/test_modeling_mamba2.py#L260).

Separately, when running `utils/check_modular_conversion.py` I get the following error:

```
Differences found between the generated code and src/transformers/models/zamba2/modeling_zamba2.py:

   1 --- src/transformers/models/zamba2/modeling_zamba2.py_generated
   2 +++ src/transformers/models/zamba2/modeling_zamba2.py
   3 @@ -313,6 +313,13 @@
   4      return attn_output, attn_weights
   5  
   6  
   7 +def rotate_half(x):
   8 +    """Rotates half the hidden dims of the input."""
   9 +    x1 = x[..., : x.shape[-1] // 2]
  10 +    x2 = x[..., x.shape[-1] // 2 :]
  11 +    return torch.cat((-x2, x1), dim=-1)
  12 +
  13 +
  14  def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
  15      """Applies Rotary Position Embedding to the query and key tensors.
  16  
  17 @@ -338,13 +345,6 @@
  18      q_embed = (q * cos) + (rotate_half(q) * sin)
  19      k_embed = (k * cos) + (rotate_half(k) * sin)
  20      return q_embed, k_embed
  21 -
  22 -
  23 -def rotate_half(x):
  24 -    """Rotates half the hidden dims of the input."""
  25 -    x1 = x[..., : x.shape[-1] // 2]
  26 -    x2 = x[..., x.shape[-1] // 2 :]
  27 -    return torch.cat((-x2, x1), dim=-1)
```
which I was not getting before despite this part was identical.

pglorio committed:
rebase

Cyrilvallez reviewed (approved):
LGTM! Let's just wait for https://github.com/huggingface/transformers/pull/35795 which will get rid of the CI failure for modular conversion! Sorry about that, and thanks for being so patient with us 🙏🙏🤗
Great work!

pglorio commented:
Awesome, sounds good!

pglorio committed:
rebase

pglorio committed:
small fix in modular file

ArthurZucker commented on docs/source/en/model_doc/zamba2.md:
````diff
@@ -0,0 +1,93 @@
+
+### Presequities
+
+Zamba2 requires you use `transformers` version 4.46.0 or higher:
+```bash
+pip install transformers>=4.46.0
+```
````
```suggestion
Zamba2 requires you use `transformers` version 4.48.0 or higher:
```bash
pip install transformers>=4.48.0
```
```

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+    return num_gs
+
+
+def layer_type_list(config: Zamba2Config):
+    """
+    Returns list of layer ids containing hybrid layers
+    """
+    output_list = []
+    for index, type in enumerate(config.layers_block_type):
+        if type == "hybrid":
+            output_list.append(index)
+    return output_list
```
I don't understand why we have this when we can simply store the explicit list in the config?

pglorio replied:
done.
added [this](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-a3e1ae24619c2d0cc1c51bfcdda1e5e647abc0d3328283db504d6c8fc0d01663R237) to the config and replaced `layer_type_list(config)` with `config.hybrid_layer_ids` in modular_zamba2.py [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a).

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+    pass
+
+
+def count_mem_blocks_in_config(config: Zamba2Config):
+    """
+    Count number of shared blocks
+    """
+    num_gs = 0
+    for val in config.layers_block_type:
+        if val == "hybrid":
+            num_gs += 1
```
same for this + it's only used once, not sure it's worth doing this

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58) and [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aR755).

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        self._modules = {}
+        self._parameters = {}
+        self._buffers = {}
+        self.conv_states = {
+            i: torch.zeros(
+                batch_size,
+                self.intermediate_size + 2 * config.mamba_ngroups * config.mamba_d_state,
+                self.conv_kernel_size,
+                device=device,
+                dtype=dtype,
+            )
+            for i in range(config.num_hidden_layers)
+        }
+        self.ssm_states = {
+            i: torch.zeros(
+                batch_size, self.n_mamba_heads, config.mamba_headdim, self.ssm_state_size, device=device, dtype=dtype
+            )
+            for i in range(config.num_hidden_layers)
+        }
+        for i in range(config.num_hidden_layers):
+            if self.layers_block_type[i] == "hybrid":
+                self.transformer_layers.append(i)
```
a single for loop should suffice here

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL132-L148)

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        self.self_attn = Zamba2Attention(config, layer_idx=-1, num_fwd_mem_blocks=num_gs, block_id=block_id)
+        self.feed_forward = Zamba2MLP(config, num_fwd_mem_blocks=num_gs, block_id=block_id)
+
+    def forward(
```
forward can be the same as `ZambaAttentionDecoderLayer` no?

ArthurZucker replied:
we can prob remove the cache_positions as they are not used in both modeling

pglorio replied:
I agree, deleted the forward [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL784-L794).

also removed `cache_position` from zamba [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-0f4d89960530c068a10af906f1958ed46e3e5f2ff937d6be61517478f383b074).

pglorio replied:
actually the forward is not the same as `ZambaAttentionDecoderLayer`. I had to revert this ([here](https://github.com/huggingface/transformers/pull/34517/commits/929ee67bc3665373887ac1e41455a182b44e958b)) and reinstate the overridden forward of Zamba2. They differ in [this line](https://github.com/huggingface/transformers/blob/929ee67bc3665373887ac1e41455a182b44e958b/src/transformers/models/zamba2/modular_zamba2.py#L802), the MLP takes an additional argument `layer_idx` in Zamba2.

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        self.gate_up_proj = nn.Linear(self.hidden_size, 2 * self.intermediate_size, bias=config.add_bias_linear)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.add_bias_linear)
+
+        if self.config.use_shared_mlp_adapter:
```
is this true of false for released checkpoint?

pglorio replied:
this is True for the released checkpoints

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        return self.torch_forward(hidden_states, cache_params, attention_mask)
+
+
+class Zamba2MLP(ZambaMLP):
```
both the init and the forward are fairly different, no point in inheriting here IMO!

pglorio replied:
agree, dropped the inheritance [here](https://github.com/huggingface/transformers/pull/34517/commits/02f58079962e1924e01844d0c9581241695d5a58#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7aL734-R723)

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        self.mamba = Zamba2MambaMixer(config=config, layer_idx=layer_idx)
+        self.input_layernorm = Zamba2RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    def forward(
```
as you are inheriting from zamba mamba decoder layer, you don't need this forward

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/bfb026750ec1277407ceda71fbe9fc873cd4e87f).

I also had to remove `position_ids` from various forward signatures in zamba to make the zamba mamba decoder forward inheritable

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        self, shared_transformer: Zamba2AttentionDecoderLayer, linear: nn.Linear, mamba: Zamba2MambaDecoderLayer
+    ):
+        super().__init__(shared_transformer, linear, mamba)
+        del self.shared_transf
```
wow I wish I caught this when reviewing the original model 🤣

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+        return layer_outputs
+
+
+ZAMBA2_START_DOCSTRING = r"""
+    This model inherits from [`PreTrainedModel`]. Check the superclass documentation for the generic methods the
+    library implements for all its model (such as downloading or saving, resizing the input embeddings, pruning heads
+    etc.)
+
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+
+    Parameters:
+        config ([`Zamba2Config`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+
+@add_start_docstrings(
+    "The bare Zamba2 Model outputting raw hidden-states without any specific head on top.",
+    ZAMBA2_START_DOCSTRING,
+)
```
pretty sure removing them can work with auto renaming!

pglorio replied:
[done](https://github.com/huggingface/transformers/pull/34517/commits/b2229430a3cae1e66f979464ddd42eda056728a5)

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+"""
+
+
+@add_start_docstrings(
+    "The bare Zamba2 Model outputting raw hidden-states without any specific head on top.",
+    ZAMBA2_START_DOCSTRING,
+)
```
same, if they are the same input as zamba, you don't need to explicitly write these

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/b2229430a3cae1e66f979464ddd42eda056728a5) too

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1388 @@
+                        "shared_transformer.pre_ff_layernorm.weight",
+                    ]
+                    self._tied_weights_keys = [*self._tied_weights_keys, *[prefix_name + key for key in tied_keys]]
+                    if self.config.use_shared_mlp_adapter:
```
same comment about code path, which models have this set to true / false?

ArthurZucker replied:
+ tied key supports regex patter, we should never have to add all of themmanually like this

pglorio replied:
all checkpoints have `config.use_shared_mlp_adapter` set to `True`. We have internal checkpoints with this flag set to `False`, which might be released in the future.

ArthurZucker replied:
I'd rather we add a new model when they are release than having 2 code pahts 😉 it's two different models for us!

pglorio replied:
replaced tied keys with regex patterns [here](https://github.com/huggingface/transformers/pull/34517/commits/87b938b488a15df501b9fe7b522c78e9ba1d26eb)

pglorio replied:
> I'd rather we add a new model when they are release than having 2 code pahts 😉 it's two different models for us!

sounds good, got rid of `config.use_shared_mlp_adapter` [here](https://github.com/huggingface/transformers/pull/34517/commits/8ed23534535afd0e17b252406878894744cb914c).

ArthurZucker commented on tests/models/zamba2/test_modeling_zamba2.py:
```diff
@@ -0,0 +1,659 @@
+            , dtype=torch.float32)  # fmt: skip
+
+        torch.testing.assert_close(logits[0, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_0, rtol=1e-3, atol=1e-3)
+        torch.testing.assert_close(logits[1, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_1, rtol=1e-3, atol=1e-3)
```
It's missing a test on cpu with the sow forward!

pglorio replied:
could you please say more about this?

pglorio replied:
done [here](https://github.com/huggingface/transformers/pull/34517/commits/5e092909971410e31d82531b718ed1ab0fe71221) for both `test_simple_generate` and `test_simple_batched_generate_with_padding`. 

`test_simple_generate` passes the cpu test straightforwardly. `test_simple_batched_generate_with_padding` marginally doesn't pass on cpu for one of the two output logits in the batch (disagreement on 2 out of 40 logits):
```
        torch.testing.assert_close(logits[0, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_0, rtol=1e-3, atol=1e-3)
>       torch.testing.assert_close(logits[1, -1, :40].cpu(), EXPECTED_LOGITS_NO_GRAD_1, rtol=1e-3, atol=1e-3)
E       AssertionError: Tensor-likes are not close!
E       
E       Mismatched elements: 2 / 40 (5.0%)
E       Greatest absolute difference: 0.009563922882080078 at index (12,) (up to 0.001 allowed)
E       Greatest relative difference: 0.030748309567570686 at index (0,) (up to 0.001 allowed)
```
Given this is a 1.2B parameter model, it's not so surprising to find occasional small discrepancies in the forward pass when running a model of this size on CPU instead of GPU. I updated the value of the absolute tolerance to a new value when the test is run on CPU [here](https://github.com/huggingface/transformers/pull/34517/commits/5e092909971410e31d82531b718ed1ab0fe71221#diff-67d87c36da1310eb4bb253eb029c63484c1a052ae06224eb5e69e72755ec46b2R665): `atol=1e-3` -> `atol=6e-3 if torch_device == "cpu" else 1e-3`.

ArthurZucker replied:
yep sounds good thanks a lot for checking this

ArthurZucker reviewed (commented):
Thanks! A few comments about the code paths, regex init and should be good!

pglorio committed:
Update docs/source/en/model_doc/zamba2.md

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

pglorio committed:
several minor fixes

pglorio committed:
inherit mamba2decoder fwd and drop position_ids in mamba

pglorio committed:
removed docstrings from modular

pglorio committed:
rebase

pglorio committed:
reinstate zamba2 attention decoder fwd

pglorio committed:
use regex for tied keys

pglorio committed:
Revert "use regex for tied keys"

This reverts commit 9007a522b1f831df6d516a281c0d3fdd20a118f5.

pglorio committed:
use regex for tied keys

pglorio committed:
add cpu to slow forward tests

pglorio committed:
dropped config.use_shared_mlp_adapter

pglorio commented:
Thank you @ArthurZucker! I think all your comments have been addressed. All zamba-related tests appear to pass!

ArthurZucker commented on docs/source/en/model_doc/zamba2.md:
````diff
@@ -0,0 +1,92 @@
+Zamba2 requires you use `transformers` version 4.48.0 or higher:
+```bash
+pip install transformers>=4.48.0
+
````
```suggestion
```
```
missing closing brackets

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1156 @@
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+        if self.config.use_shared_attention_adapter:
```
I don'tknow if I asked already, but similarly is this true / false for the released checkpoint?

pglorio replied:
this is true for some of the released checkpoints and false for other released checkpoints

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1156 @@
+        key_states = key_states.view(hidden_shape).transpose(1, 2)
+        value_states = value_states.view(hidden_shape).transpose(1, 2)
+
+        if self.config.use_mem_rope:
```
same comment, let's weed out the final bits that are not part of the released checkpoint!

ArthurZucker commented on src/transformers/models/zamba2/modular_zamba2.py:
```diff
@@ -0,0 +1,1156 @@
+
+        self._attn_implementation = config._attn_implementation
+        self.final_layernorm = Zamba2RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        if config.use_mem_rope:
+            if config.use_long_context:
+                logger.warning_once(
+                    "`use_long_context` set to `True`: using rescaled `rope_theta` and extended `max_position_embeddings`."
+                )
```
same comment here!

pglorio replied:
same for this, this is a flag that changes theta and increases model's performance at long-context tasks for the 7B checkpoint, so it can be both true and false

ArthurZucker reviewed (approved):
Thanks only a small comment left regarding code paths and good to go !

pglorio committed:
Update docs/source/en/model_doc/zamba2.md

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

pglorio commented:
Thanks @ArthurZucker, I replied to your comments above.

pglorio committed:
rebase

pglorio committed:
re-convert from modular
